src/D2L.CodeStyle.Analyzers/ApiUsage/IdTypesAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class IdTypesAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.IdTypeParameterMismatch
+		);
+
+		private static readonly ImmutableDictionary<string, ImmutableHashSet<string>> PARAM_TO_VARIABLE_BLACKLIST = new Dictionary<string, string[]> {

```

Comment:
What if it was still `string -> [String]` but instead of mapping to a blacklist we map to synonyms, like this:

```cs
ImmutableDictionary<string, ImmutableHashSet<string>> idTypes = 
  new Dictionary<string, string[]> {
    { "userId", new [] { "user" } },
    { "orgId", new [] { "org" } },
    { "orgUnitId", new [] { "orgUnit", "course", "courseId", "semester", "department", "dept" } }, // I think some of these are used...
  }.ToImmutableDictionary(
    keySelector: x => x,
    keyComparer: StringComparer.InvariantCultureIgnoreCase,
    elementSelector: syn => syn.ToImmutableHashSet( StringComparer.InvariantCultureIgnoreCase )
  );
```

Then you could do

```cs
ImmutableArray<string> allIdLikeNames = IdTypes.Keys.Union(
  IdTypes.Values.SelectMany( x => x )
).ToImmutableArray();
```

and then finally

```cs
PARAM_TO_VARIABLE_BLACKLIST = idTypes.Keys
  .ToImmutableDictionary(
    keySelector: x => x,
    keyComparer: StringComparer.InvariantCultureIgnoreCase,
    elementSelector: canonicalIdName => allIdLikeNames
      .Where( idLikeName =>
        canonicalIdName != idLikeName
        && !idTypes[canonicalIdName].contains( idLikeName )
      )
  );
```
  
whew maybe I answered my own question